core/controllers/user_query.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2016 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Controller for user query related pages and handlers."""
+
+from core.controllers import base
+from core.domain import config_domain
+from core.domain import user_query_services
+from core.domain import user_query_jobs_one_off
+from core.domain import user_services
+from core.platform import models
+
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+
+current_user_services = models.Registry.import_current_user_services()
+
+def require_valid_sender(handler):
+    """Decorator that checks if the current user is a authorized sender."""
+    def test_user(self, **kwargs):
+        """Checks if the user is logged in and is authorized sender."""
+        if not self.user_id:
+            self.redirect(
+                current_user_services.create_login_url(self.request.uri))
+            return
+        if self.username not in config_domain.WHITELISTED_EMAIL_SENDERS.value:
+            raise self.UnauthorizedUserException(
+                '%s is not a authorized user of this application', self.user_id)
+        return handler(self, **kwargs)
+
+    return test_user
+
+
+class QueryPage(base.BaseHandler):
+    """Page to submit query and show past queries."""
+
+    @require_valid_sender
+    def get(self):
+        """Handles GET requests."""
+        self.render_template('pages/query/query.html')
+
+
+class QueryDataHandler(base.BaseHandler):
+    """Query data handler."""
+
+    @require_valid_sender
+    def get(self):
+        offset = (
```

Comment:
Instead of manually handling offsets, how about using GAE's built-in functionality for query cursors (e.g. with fetch_page())? This is also consistent with how we handle things like displaying search results, and avoids unnecessary bookkeeping.

https://cloud.google.com/appengine/docs/python/ndb/queries#cursors